services/brig/src/Brig/Data/MLS/KeyPackage.hs

```diff
@@ -194,9 +195,47 @@ addKeyPackageRef ref nkpr = do
     q :: PrepQuery W (ClientId, ConvId, Domain, Domain, UserId, KeyPackageRef) x
     q = "UPDATE mls_key_package_refs SET client = ?, conv = ?, conv_domain = ?, domain = ?, user = ? WHERE ref = ?"
 
+-- | Update key package ref, used in Galley when commit reveals key package ref update for the sender.
+-- Nothing is changed if the previous key package ref is not found in the table.
+-- Updating amounts to INSERT the new key package ref, followed by DELETE the
+-- previous one.
+--
+-- FUTUREWORK: this function has to be extended if a table mapping (client,
+-- conversation) to key package ref is added, for instance, when implementing
+-- external delete proposals.
+updateKeyPackageRef :: MonadClient m => KeyPackageRef -> KeyPackageRef -> m ()
+updateKeyPackageRef prevRef newRef =
+  void . runMaybeT $
+    backupKeyPackageMeta prevRef
+      >>= ($) lift . ((>> deleteKeyPackage prevRef) . restoreKeyPackageMeta newRef)
+
 --------------------------------------------------------------------------------
 -- Utilities
 
+backupKeyPackageMeta :: MonadClient m => KeyPackageRef -> MaybeT m (ClientId, ConvId, Domain, Domain, UserId)
```

Comment:
Can you regroup the return type so `Qualified` is used where appropriate? Seeing two occurrences of one and the same type (`Domain` in this case) makes me uneasy as it's really easy to confuse values of the same type.

Reply:
Done. See [ea33455](https://github.com/wireapp/wire-server/pull/2604/commits/ea334556b017d02bd9643cd89caeb041e78ccce0).